cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,916 @@
+
+  __device__ thrust::tuple<vertex_t, vertex_t> operator()(edge_t i) const
+  {
+    // edges are sorted with destination as key so reverse the edge when looking it
+    auto pair = thrust::make_tuple(thrust::get<1>(*(edge_unrolled + i)),
+                                   thrust::get<0>(*(edge_unrolled + i)));
+    // Find its position in either partition of the transposed edgelist
+    // An edge can be in found in either of the two partitions (valid or invalid)
+
+    auto itr = thrust::lower_bound(
+      thrust::seq, transposed_valid_edge_last, transposed_invalid_edge_last, pair);
+    auto idx = thrust::distance(transposed_valid_edge_last, itr) + num_valid_edges;
+
+    if (*itr != pair) {
+      // The edge must be in the first boundary
+      itr = thrust::lower_bound(
+        thrust::seq, transposed_valid_edge_first, transposed_valid_edge_last, pair);
+      idx = thrust::distance(transposed_valid_edge_first, itr);
+    }
+
+    assert(*itr == pair);
+    cuda::atomic_ref<edge_t, cuda::thread_scope_device> atomic_counter(num_triangles[idx]);
+    auto r = atomic_counter.fetch_sub(edge_t{1}, cuda::std::memory_order_relaxed);
+  }
+};
+
+template <typename vertex_t, typename edge_t>
+rmm::device_uvector<vertex_t> prefix_sum_valid_and_invalid_edges(
+  // The edgelist is segmented into 2 partitions (valid and invalid edges)
+  // and edges to be unrolled can be either in the valid or the invalid edge
+  // partition.
+  raft::handle_t const& handle,
+  edge_t num_edges,
+  raft::device_span<vertex_t const> invalid_dst,
+  raft::device_span<vertex_t> edgelist_dsts)
+{
+  rmm::device_uvector<vertex_t> prefix_sum(invalid_dst.size() + 1, handle.get_stream());
+
+  thrust::tabulate(
+    handle.get_thrust_policy(),
+    prefix_sum.begin(),
+    prefix_sum.begin() + invalid_dst.size(),
+    [num_edges, invalid_dst, edgelist_dsts = edgelist_dsts.begin()] __device__(auto idx) {
+      auto itr_lower_valid = thrust::lower_bound(
+        thrust::seq, edgelist_dsts, edgelist_dsts + num_edges, invalid_dst[idx]);
+      
+      auto itr_upper_valid = thrust::upper_bound(
+        thrust::seq, itr_lower_valid, edgelist_dsts + num_edges, invalid_dst[idx]);
+
+      auto dist = thrust::distance(itr_lower_valid, itr_upper_valid);
+
+      return dist;
+    });
+  thrust::exclusive_scan(
+    handle.get_thrust_policy(), prefix_sum.begin(), prefix_sum.end(), prefix_sum.begin());
+
+  return prefix_sum;
+}
+
+template <typename vertex_t, typename edge_t, typename EdgeIterator>
+edge_t remove_overcompensating_edges(raft::handle_t const& handle,
+                                     edge_t num_valid_edges,
+                                     edge_t num_invalid_edges,
+                                     edge_t dataframe_buffer_size,
+                                     EdgeIterator&& potential_closing_or_incoming_edges,
+                                     EdgeIterator&& incoming_or_potential_closing_edges,
```

Comment:
EdgeIterator is a misnomer here. EdgeBuffer might be a better name.

Reply:
And if you're taking an R-value, it's life time should finish in this function. If not you should return those back. In the caller site, once `std::move`ed, it shouldn't be accessed again. return std::tuple<EdgeBuffer, EdgeBuffer> instead of edge_t.

Reply:
Right. I don't actually need to take the R-value so what I did instead was to pass a pointer to the beginning of the dataframe buffer.